cloud-storage/cloud_storage.c

```diff
@@ -0,0 +1,947 @@
+      struct
+      {
+         folder_type_t folder_type;
+      } sync;
+
+      struct
+      {
+         folder_type_t folder_type;
+         char *local_file;
+      } upload;
+
+      struct
+      {
+         folder_type_t folder_type;
+         cloud_storage_item_t *remote_file;
+      } download;
+   } parameters;
+};
+
+static bool _shutdown = true;
+static cloud_storage_provider_t *_provider;
+
+static char _game_states_config_value[8192] = "";
+
+static bool _need_sync_runtime_logs = false;
+
+cloud_storage_item_t *_game_states = NULL;
+
+static generic_queue_t *_operation_queue = NULL;
+static slock_t *_operation_queue_mutex;
+
+static scond_t *_operation_condition;
+static sthread_t *_operation_thread;
+
+static void _sync_files_execute(folder_type_t folder_type);
+
+static void _upload_file_execute(folder_type_t folder_type, char *file_name);
+
+void _cloud_storage_item_free_fn(void *item)
+{
+   cloud_storage_item_free((cloud_storage_item_t *) item);
+}
+
+void cloud_storage_item_free(cloud_storage_item_t *item)
+{
+   while (item)
+   {
+      free(item->id);
+      free(item->name);
+
+      if (item->item_type == CLOUD_STORAGE_FILE)
+      {
+         if (item->type_data.file.hash_value)
+         {
+            free(item->type_data.file.hash_value);
+         }
+         if (item->type_data.file.download_url)
+         {
+            free(item->type_data.file.download_url);
+         }
+      }
+
+      if (item->item_type == CLOUD_STORAGE_FOLDER)
+      {
+         linked_list_free(item->type_data.folder.children, &_cloud_storage_item_free_fn);
+      }
+
+      free(item);
+   }
+}
+
+static void _free_operation(struct _operation_t *operation)
+{
+   switch (operation->operation_type)
+   {
+      case UPLOAD_FILE:
+         if (operation->parameters.upload.local_file)
+         {
+            free(operation->parameters.upload.local_file);
+         }
+         break;
+      case DOWNLOAD_FILE:
+         if (operation->parameters.download.remote_file)
+         {
+            cloud_storage_item_free(operation->parameters.download.remote_file);
+         }
+         break;
+      default:
+         break;
+   }
+
+   free(operation);
+}
+
+static void _free_operation_fn(void *operation)
+{
+   _free_operation((struct _operation_t *) operation);
+}
+
+static void _operation_thread_loop(void *user_data)
+{
+   while (!_shutdown)
+   {
+      struct _operation_t *operation;
+
+      /* In the critical section:
+       * - Go back to waiting if the opertion queue is empty
+       * - Break out of the loop and shutdown the operation thread if
+       *   _shutdown is true
+       * - Pop an operation off of the queue and release the lock
+       */
+      slock_lock(_operation_queue_mutex);
+      if (generic_queue_length(_operation_queue) == 0)
+      {
+         scond_wait(_operation_condition, _operation_queue_mutex);
+      }
+
+      if (_shutdown)
+      {
+         slock_unlock(_operation_queue_mutex);
+         continue;
```

Comment:
Why continue? I mean it will indeed re-evaluate the while condition and stop iterating, but break is more readable right? Makes it clearer that you are aborting the loop.

Reply:
Updated.